// @egjs/ngx-view360/esm2020/lib/ngx-view360.module.mjs
/**
 * Copyright (c) 2023-present NAVER Corp.
 * egjs projects are licensed under the MIT license
 */
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { NgxView360Component } from "./ngx-view360.component";
import * as i0 from "@angular/core";
export class NgxView360Module {
}
NgxView360Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: NgxView360Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxView360Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.9", ngImport: i0, type: NgxView360Module, declarations: [NgxView360Component], imports: [CommonModule], exports: [NgxView360Component] });
NgxView360Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: NgxView360Module, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: NgxView360Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxView360Component],
                    imports: [
                        CommonModule
                    ],
                    exports: [NgxView360Component]
                }]
        }] });
